lambda_code/src/models/fhir_immunization.py

```diff
@@ -754,6 +754,389 @@ def pre_validate_reason_code_coding_displays(cls, values: dict) -> dict:
+        NHS_NUMBER_STATUS_INDICATOR_CODE) exists, then it is a non-empty string
+        """
+        try:
+            nhs_number_status_code = get_generic_questionnaire_response_value(
+                values, "NhsNumberStatus", "code"
+            )
+            pre_validate_string(
+                nhs_number_status_code,
+                generate_field_location_for_questionnnaire_response(
+                    link_id="NhsNumberStatus", field_type="code"
+                ),
+            )
+        except KeyError:
+            pass
+
+        return values
+
+    @classmethod
+    def pre_validate_nhs_number_status_display(cls, values: dict) -> dict:
+        """
+        Pre-validate that, if contained[?(@.resourceType=='QuestionnaireResponse')]
+        .item[?(@.linkId=='NhsNumberStatus')].answer[0].valueCoding.display (legacy CSV field name:
+        NHS_NUMBER_STATUS_INDICATOR_DESCRIPTION) exists, then it is a non-empty string
+        """
+        try:
+            nhs_number_status_display = get_generic_questionnaire_response_value(
+                values, "NhsNumberStatus", "display"
+            )
+            pre_validate_string(
+                nhs_number_status_display,
+                generate_field_location_for_questionnnaire_response(
+                    link_id="NhsNumberStatus", field_type="display"
+                ),
+            )
+        except KeyError:
+            pass
+
+        return values
+
+    @classmethod
+    def pre_validate_site_code_system(cls, values: dict) -> dict:
+        """
+        Pre-validate that, if contained[?(@.resourceType=='QuestionnaireResponse')]
+        .item[?(@.linkId=='SiteCode')].answer[0].valueCoding.system (legacy CSV field name:
+        SITE_CODE_TYPE_URI) exists, then it is a non-empty string
+        """
+        try:
+            site_code_system = get_generic_questionnaire_response_value(
+                values, "SiteCode", "system"
+            )
+            pre_validate_string(
+                site_code_system,
+                generate_field_location_for_questionnnaire_response(
+                    link_id="SiteCode", field_type="system"
+                ),
+            )
+        except KeyError:
+            pass
+
+        return values
+
+    @classmethod
+    def pre_validate_local_patient_code(cls, values: dict) -> dict:
+        """
+        Pre-validate that, if contained[?(@.resourceType=='QuestionnaireResponse')]
+        .item[?(@.linkId=='LocalPatient')].answer[0].valueCoding.code (legacy CSV field name:
+        LOCAL_PATIENT_ID) exists, then it is a non-empty string
+        """
+        try:
+            local_patient_code = get_generic_questionnaire_response_value(
+                values, "LocalPatient", "code"
+            )
+            pre_validate_string(
+                local_patient_code,
+                generate_field_location_for_questionnnaire_response(
+                    link_id="LocalPatient", field_type="code"
```

Comment:
Also needs to handle 'max 20 characters rule' (see word document FHIR API spec)

Reply:
done